### PR TITLE
[IIDM v1.2] Add convenient methods to check tap changers' presence on transformers (#131)

### PR DESCRIPTION
**Pull request by sebalaig**
Backport of https://github.com/powsybl/powsybl-core/pull/1346 and https://github.com/powsybl/powsybl-core/pull/1391.

Signed-off-by: Sébastien LAIGRE <slaigre@silicom.fr>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*
#131 

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change or deprecate an API?** *If yes, check the following:*
- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: include/powsybl/iidm/PhaseTapChangerHolder.hpp [hunk 1]
```diff
@@ -22,10 +22,16 @@ class PhaseTapChangerHolder : public virtual TapChangerHolder {
...
 
+    stdcxx::CReference<PhaseTapChanger> getOptionalPhaseTapChanger() const;
+
+    stdcxx::Reference<PhaseTapChanger> getOptionalPhaseTapChanger();
+
     virtual stdcxx::CReference<PhaseTapChanger> getPhaseTapChanger() const = 0;
 
     virtual stdcxx::Reference<PhaseTapChanger> getPhaseTapChanger() = 0;
 
+    bool hasPhaseTapChanger() const override;
```
Comment (mathbagu) [inline include/powsybl/iidm/PhaseTapChangerHolder.hpp:33]: From a user point of view, it's really strange to have both `getOptionalPhaseTapChanger` and `getPhaseTapChanger` that return the same value. In java, we introduce the optional version for functional programming purpose.
2 options:
- only introduce the `hasPhaseTapChanger` method
- introduce optional methods, but change the return type of `getPhaseTapChanger` to return a reference (and throw an exception if not set even if it's different from the java version).

@gautierbureau What do you think?

Comment (gautierbureau) [reply include/powsybl/iidm/PhaseTapChangerHolder.hpp:33]: Personally I prefer first solution

Comment (mathbagu) [reply include/powsybl/iidm/PhaseTapChangerHolder.hpp:33]: I would choose the 2nd one :D

Comment (sebalaig) [reply include/powsybl/iidm/PhaseTapChangerHolder.hpp]: Updated in b064c3adbb2481d1634bff763453cb96fb448451 to implement the second one.

File: include/powsybl/iidm/RatioTapChangerHolder.hpp [hunk 2]
```diff
@@ -22,9 +22,13 @@ class RatioTapChangerHolder : public virtual TapChangerHolder {
 public:
     ~RatioTapChangerHolder() noexcept override = default;
 
-    virtual stdcxx::CReference<RatioTapChanger> getRatioTapChanger() const = 0;
+    virtual stdcxx::CReference<RatioTapChanger> getOptionalRatioTapChanger() const;
 
-    virtual stdcxx::Reference<RatioTapChanger> getRatioTapChanger() = 0;
+    virtual stdcxx::Reference<RatioTapChanger> getOptionalRatioTapChanger();
```
Comment (mathbagu) [inline include/powsybl/iidm/RatioTapChangerHolder.hpp:27]: Both optional methods should be declare as pure virtual

File: include/powsybl/iidm/PhaseTapChangerHolder.hpp [hunk 3]
```diff
@@ -22,9 +22,13 @@ class PhaseTapChangerHolder : public virtual TapChangerHolder {
...
 
-    virtual stdcxx::CReference<PhaseTapChanger> getPhaseTapChanger() const = 0;
+    virtual stdcxx::CReference<PhaseTapChanger> getOptionalPhaseTapChanger() const;
 
-    virtual stdcxx::Reference<PhaseTapChanger> getPhaseTapChanger() = 0;
+    virtual stdcxx::Reference<PhaseTapChanger> getOptionalPhaseTapChanger();
+
+    virtual const PhaseTapChanger& getPhaseTapChanger() const = 0;
+
+    virtual PhaseTapChanger& getPhaseTapChanger() = 0;
```
Comment (mathbagu) [inline include/powsybl/iidm/PhaseTapChangerHolder.hpp:31]: Why hasPhaseTapChanger is not declared in this interface?

File: include/powsybl/iidm/RatioTapChangerHolder.hpp [hunk 4]
```diff
@@ -22,9 +22,13 @@ class RatioTapChangerHolder : public virtual TapChangerHolder {
...
+
+    virtual const RatioTapChanger& getRatioTapChanger() const = 0;
+
+    virtual RatioTapChanger& getRatioTapChanger() = 0;
```
Comment (mathbagu) [inline include/powsybl/iidm/RatioTapChangerHolder.hpp:31]: Why hasRatioTapChanger is not declared in this interface?

File: src/iidm/ThreeWindingsTransformer.cpp [hunk 5]
```diff
@@ -129,11 +141,11 @@ double ThreeWindingsTransformer::Leg::getX() const {
...
 
 bool ThreeWindingsTransformer::Leg::hasPhaseTapChanger() const {
-    return static_cast<bool>(m_phaseTapChanger);
+    return m_phaseTapChanger != nullptr;
 }
 
 bool ThreeWindingsTransformer::Leg::hasRatioTapChanger() const {
-    return static_cast<bool>(m_ratioTapChanger);
+    return m_ratioTapChanger != nullptr;
 }
```
Comment (mathbagu) [inline src/iidm/ThreeWindingsTransformer.cpp:149]: Not really efficient: `nullptr` will be implicitly converted to a `unique_ptr` whereas a `static_cast<bool>()` seems better.

File: src/iidm/TwoWindingsTransformer.cpp [hunk 6]
```diff
@@ -137,6 +141,14 @@ double TwoWindingsTransformer::getX() const {
...
 }
 
+bool TwoWindingsTransformer::hasPhaseTapChanger() const {
+    return m_phaseTapChanger != nullptr;
+}
+
+bool TwoWindingsTransformer::hasRatioTapChanger() const {
+    return m_ratioTapChanger != nullptr;
+}
+
```
Comment (mathbagu) [inline src/iidm/TwoWindingsTransformer.cpp:151]: Same remark

File: include/powsybl/iidm/PhaseTapChangerHolder.hpp (hunk 3 above)
Comment (sebalaig) [reply include/powsybl/iidm/PhaseTapChangerHolder.hpp:31]: It is inherited from `TapChangerHolder` and still virtual pure in `RatioTapChangerHolder`/`PhaseTapChangerHolder`. Do I have to redeclare it in both subclasses `RatioTapChangerHolder`/`PhaseTapChangerHolder` ?

File: include/powsybl/iidm/RatioTapChangerHolder.hpp (hunk 2 above)
Comment (mathbagu) [reply include/powsybl/iidm/RatioTapChangerHolder.hpp:27]: These methods should not be declared as virtual

**Pull request merged**